Remove deprecated stochastic swap transpiler pass

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit removes the stochastic swap transpiler pass and its associated routing stage plugin. This pass has been superseded by sabre at this point and it was deprecated in the Qiskit 1.3.0 release.

### Details and comments

Part of #13699

mtreinish committed:
Remove deprecated stochastic swap transpiler pass

This commit removes the stochastic swap transpiler pass and its
associated routing stage plugin. This pass has been superseded by sabre
at this point and it was deprecated in the Qiskit 1.3.0 release.

Part of #13699

jakelishman commented on releasenotes/notes/remove-stochastic-swap-88e2d6be4c0d5713.yaml:
```diff
@@ -0,0 +1,42 @@
+          target=backend.target
+      )
+
+      this should be replaced with::
```
Need to unindent this text line to separate the code blocks.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13791/commits/1c0efbd7d66bcfe52111650409cefe26f50c5aa4

jakelishman commented on releasenotes/notes/remove-stochastic-swap-88e2d6be4c0d5713.yaml:
```diff
@@ -0,0 +1,42 @@
+upgrade_transpiler:
+  - |
+    The deprecated ``StochasticSwap`` transpiler pass, and it's associated
+    built-in routing stage plugin `"stochastic"` have been removed. These
```
```suggestion
    The deprecated ``StochasticSwap`` transpiler pass, and its associated
    built-in routing stage plugin `"stochastic"`, have been removed. These
```

jakelishman commented on test/python/transpiler/test_mappers.py:
```diff
@@ -106,12 +106,6 @@ def create_passmanager(self, coupling_map, initial_layout=None):
-                message=r".*StochasticSwap.*",
-            )
             # pylint: disable=not-callable
             passmanager.append(self.pass_class(CouplingMap(coupling_map), **self.additional_args))
```
Can remove the whole `warnings.catch_warnings` now too

mtreinish replied:
Fixed in https://github.com/Qiskit/qiskit/pull/13791/commits/1063f06284bdbb7fc41a1b1f3272cef7d0c5c1d5

jakelishman commented on test/python/transpiler/test_sabre_layout.py:
```diff
@@ -308,12 +307,10 @@ def test_support_var_with_explicit_routing_pass(self):
-                cm, seed=0, routing_pass=StochasticSwap(cm, trials=1, seed=0, fake_run=True)
-            )
+            pass_ = SabreLayout(cm, seed=0, routing_pass=BasicSwap(cm, fake_run=True))
             _ = pass_(qc)
```
This _shouldn't_ warn now... I think it might be the duration thing inside the implicit `dag_to_circuit` from `SabreLayout.__call__`.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13791/commits/1063f06284bdbb7fc41a1b1f3272cef7d0c5c1d5

jakelishman reviewed (commented):
o7 to the first Rust code of Qiskit

mtreinish committed:
Update releasenotes/notes/remove-stochastic-swap-88e2d6be4c0d5713.yaml

Co-authored-by: Jake Lishman <jake@binhbar.com>

jakelishman commented:
Oh also I meant to write in the top review comment: we need to update the `qiskit.transpiler` and `qiskit.transpiler.preset_passmanager.plugin` docstrings to remove the reference to the `stochastic` plugin.

jakelishman commented:
There's also a couple of places in the benchmark suite that refer to the `stochastic` plugin.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into farewell-old-friend

mtreinish committed:
Remove stochastic swap from benchmarks

mtreinish committed:
Remove stochastic swap plugin details from transpiler docs

mtreinish committed:
Update tests

mtreinish committed:
Fix release note formatting

mtreinish committed:
Merge branch 'main' into farewell-old-friend

mtreinish commented:
> There's also a couple of places in the benchmark suite that refer to the stochastic plugin.

Fixed in: https://github.com/Qiskit/qiskit/pull/13791/commits/b057980522d8684a67c51bc3df782a76af799c91

> Oh also I meant to write in the top review comment: we need to update the qiskit.transpiler and qiskit.transpiler.preset_passmanager.plugin docstrings to remove the reference to the stochastic plugin.

Fixed in https://github.com/Qiskit/qiskit/pull/13791/commits/cdf7826caa466dc7553a6d86bcd9b4b6a57d5230

I opted to keep the reference I found in `plugin.py` because it looked like the only mention there was that "stochastic" was marking it as a reserved which seems fine to keep just so people don't have issues if they want to develop a plugin that supports 1.x and 2.x.

jakelishman reviewed (dismissed):
I can live with leaving `stochastic` in the "reserved names" list, though it wouldn't surprise me if there's a PR in 6 months' time deleting it, confused about why we're reserving something we don't use haha.

jakelishman commented:
Oh, a lint complaint about `import warnings` in `test/python/transpiler/test_mappers.py`.

mtreinish committed:
Remove unused import

mtreinish committed:
Merge branch 'main' into farewell-old-friend

mtreinish committed:
Merge branch 'main' into farewell-old-friend
